Create comment page refactor

hjiangsu opened the pull request:
## Pull Request Description

This PR refactors the existing create comment page and attempts to solve the following issues:
- It updates the create comment page to match logic similar to the create post page. This means that draft logic is all contained within the create comment page. It also means that all logic related to comment creation is contained within the page itself. This allows us to perform comment actions from anywhere without relying on passing in the correct blocs.
  - When a comment is created or edited successfully, it calls a callback function `onCommentSuccess`. The parent widget can use this callback function to update the state of the parent (e.g., add comment to list, update comment content for edits)
- Adds the ability to select a language for a comment. This brings us up to parity with the create post page.
- Updated the UI/UX to better match the overall UI of Thunder.
  - When replying to a post, the post information will be displayed in a similar fashion to the post page. It uses the pre-existing `PostSubview` widget, so any changes to the `PostSubview` will reflect on this page!
  - When replying to a comment, the comment information will be displayed in a similar fashion to the comment cards. It uses the pre-existing `CommentContent` to display the comment, so any changes there will also reflect on this page!

Most of the logic is now in place, and should be more or less functional! Some standing issues to be fixed:
- Need to add logic to update comment from the account page

Additional notes:
- I've disabled the comment action on the search page, and on the user feed page. My main reasoning here is that I'd like for users to visit the comment within it's full context in order to perform actions on it (vote, reply, etc). This also simplifies some of the overall logic and reduces edge cases! Do let me know your thoughts on this change though.

<!--- Please describe what was changed -->

## Issue Being Fixed

<!-- Please describe the problem that is being fixed and, if applicable, reference a GitHub issue -->

Issue Number: #1012, #738

## Screenshots / Recordings

<!-- This section is optional but highly recommended to show off your changes! -->

| Before | After |
|--------|--------|
| ![simulator_screenshot_11035348-06C9-4B12-BE0B-BEEFBE4E4E40](https://github.com/thunder-app/thunder/assets/30667958/0df6d532-4f21-49b0-83e3-389d6b4ab0e0) | ![simulator_screenshot_44F7F5FD-9615-4620-9B43-32BCE9C7407B](https://github.com/thunder-app/thunder/assets/30667958/c409a225-3d70-4b46-99b3-52bb2c12f2b8) |
| ![simulator_screenshot_04CEDB6F-B77B-47A1-9AF5-457395972889](https://github.com/thunder-app/thunder/assets/30667958/a1a0d7c6-8f40-4765-9b28-a1effa09e073) | ![simulator_screenshot_71311E3F-11F5-447E-94F5-F703E6837840](https://github.com/thunder-app/thunder/assets/30667958/a3497813-f803-4e09-aabd-abf78b8e4f42) |
| ![simulator_screenshot_82411B07-723F-4D80-8BC1-618C4382C51E](https://github.com/thunder-app/thunder/assets/30667958/a67fbc06-4cf2-47cd-a6ad-5d265e8f4942) | ![simulator_screenshot_A097A0DD-4BA7-4068-AE4D-9B1E1E455953](https://github.com/thunder-app/thunder/assets/30667958/5e19c01b-8203-420a-bfbf-c8817b281e89) | 
| ![simulator_screenshot_B5B4150B-7D1D-460C-BF91-E161712B17AA](https://github.com/thunder-app/thunder/assets/30667958/559519b7-979c-478f-9e77-8507088f3df2) | ![simulator_screenshot_7112944A-DF17-48AF-94B4-29F382A1F2FD](https://github.com/thunder-app/thunder/assets/30667958/a4b8e6b7-a3e0-4a17-8b02-8a0c6ea39a66) | 

## Checklist

- [ ] Did you update CHANGELOG.md?
- [ ] Did you use localized strings where applicable?
- [ ] Did you add `semanticLabel`s where applicable for accessibility?

micahmo commented:
Just an FYI, I want to work on a "comment as other user" feature as some point. Let me know if you think I should wait for this PR or go ahead now.

hjiangsu commented:
I think it might be good to hold off on that until this is complete! I'll probably get back on this soon

hjiangsu commented:
@micahmo this should be more or less ready now (with the exception of the standing issues). I've updated the description to provide some more information/context. This is a relatively large change, so up to you if you want to do a code review on this!

@machinaeZER0 since I've updated the overall UI/UX, please let me know your thoughts on this!

machinaeZER0 commented:
Hello! This is looking really nice :) Just a few thoughts/clarifications from me:

- I definitely see the usefulness of a language setting given the issues its absence is causing - I do wonder if there's a classier way to integrate it, though. are we able to default the language in a given community/instance to something acceptable, or no? And what are your thoughts on making the language selector a dropdown icon next to your personal info, rather than a text dropdown between your info and the composition section? (mostly unsure of what the icon would be - maybe just a three dot guy?)
![image](https://github.com/thunder-app/thunder/assets/3624023/4ad8eeae-618d-4e6d-9573-2d7cdc672e64)

- I think this is naturally improved having switched from an outline to a filled box style, but I think the quoted text isn't vertically centered. If that's true (might be my eyes playing tricks on me), is it possible to add a few pixels of padding to the bottom? 
![image](https://github.com/thunder-app/thunder/assets/3624023/b1d44310-79e0-4554-b895-46bff877f2fa)

- I'm sure this is outside the scope of this PR, but I still wish the send/post button was closer to the bottom, lolz. I think I'm in the minority on that one, but I like where it lives in most messaging apps from a reachability standpoint, rather than at the top of the screen. Maybe something to ponder for the future, but happy to keep things as they are on that front for the foreseeable :)

Let me know if I missed anything else you specifically wanted reviewed, but hope this was helpful!

micahmo commented:
Hey all, I'm still reviewing this PR, but I wanted to respond to @machinaeZER0's comments real quick.

> * are we able to default the language in a given community/instance to something acceptable

Given that the Lemmy UI does not, we probably wouldn't either.

| ![image](https://github.com/thunder-app/thunder/assets/7417301/e3d4f322-666c-4f6d-b8d0-8727c304b3d3) |
| - |

> * what are your thoughts on making the language selector a dropdown icon next to your personal info

As long as it's obviously a language selector, I'm totally up for UI suggestions. Just keep in mind that this currently matches the Create Post page, so we'd have to change both. Also keep in mind that this page will also eventually have the "comment as other user..." feature, so there will be a chevron to the right of the username (where you circled) indicating you can pick a different user.

> * I still wish the send/post button was closer to the bottom

I also agree with this! A while back I did back I tried a mockup where the send button integrated with the preview button. I'll have to dig that up. Again, that would be a common change between this and the Create Post page, so probably a different PR as you mentioned.

machinaeZER0 commented:
That all makes sense to me - if we're matching the "normal" Lemmy experience, thoughts on matching the language and styling of the dropdown here? A single dropdown (maybe pilled) that says "Select Language," rather than the current "Language: Select Language," I mean
![image](https://github.com/thunder-app/thunder/assets/3624023/1945826e-5047-4b4e-8ead-677b83efc66e)

I think it being shorter and contained within a pill would be positive changes, but also wouldn't be opposed to just doing the latter and keeping the longer copy of the former :)

machinaeZER0 commented:
Made issue #1309 as a place to discuss the send/post button idea, also :)

micahmo commented on lib/comment/view/create_comment_page.dart:
```diff
@@ -0,0 +1,561 @@
+                                  MarkdownType.separator,
+                                  MarkdownType.code,
+                                  MarkdownType.username,
+                                  MarkdownType.community,
```
Did we lose spoiler here?

hjiangsu replied:
Done!

micahmo commented on lib/comment/view/create_comment_page.dart:
```diff
@@ -0,0 +1,561 @@
+///
+/// Passing in [languageId] will set the initial state of the widget to display that given language.
+/// A callback function [onLanguageSelected] will be triggered whenever a new language is selected from the dropdown.
+class LanguageSelector extends StatefulWidget {
```
Doesn't have to be now, but is there any way to share this with the similar create post widget?

hjiangsu replied:
I think we'll make this a separate PR, so that we can apply the suggestions that @machinaeZER0 mentioned with regards to the select language input!

micahmo commented on lib/shared/snackbar.dart:
```diff
@@ -28,30 +28,33 @@ void showSnackbar(
-                        trailingIcon,
-                        color: trailingIconColor ?? Theme.of(context).colorScheme.inversePrimary,
+            content: GestureDetector(
+              onTap: () => OverlaySupportEntry.of(context)?.dismiss(),
```
Does this dismiss snackbars on tap? I wonder if we should have an X button for that.

<img width="1024" alt="unnamed" src="https://github.com/thunder-app/thunder/assets/7417301/088ce96e-d0d6-427d-b9c4-a53897df6cc9">

https://m3.material.io/components/snackbar/specs

hjiangsu replied:
Good call!

I'll switch the approach here - the original intention was to have a way to dismiss the snackbar without swiping (while having an action present). This was because the "restored comment from draft" snackbar was annoying, and blocked the markdown toolbar entirely.

hjiangsu replied:
Done in https://github.com/thunder-app/thunder/pull/1165/commits/c4fb8c4d07363d5a32b336a72481c429ab749e5e!

micahmo reviewed (approved):
Awesome work, I know this was a huge effort! I did review the code but haven't tested yet (I'm ok with getting this merged in to test).

> I've disabled the comment action on the search page, and on the user feed page.

Does this also need to be done for inbox replies and mentions (the latter of which uses a totally different UI, not swipe gestures)?

hjiangsu commented:
Thanks for the feedback @machinaeZER0!

> are we able to default the language in a given community/instance to something acceptable, or no?

As @micahmo mentioned, we're mostly following what Lemmy UI does. I do believe we can make slight improvements to the language selector in future PRs (such as limiting the language options based on what is enabled on the instance/community).

Selecting a default language would be difficult because instances/communities can have more than one language. However, if the instance/community only has one language enabled, I can see making that the default!

> And what are your thoughts on making the language selector a dropdown icon next to your personal info, rather than a text dropdown between your info and the composition section?

As @micahmo mentioned,  the only issue here is that we'll be implementing ways to choose a different user to post the comment as (in the future). This would likely follow similar UI patterns to how the create post page looks like (with the chevron icon)

> is it possible to add a few pixels of padding to the bottom?

I can try to adjust that a bit more - it could be because of the spacing of the title text which causes an optical illusion

hjiangsu commented:
> Does this also need to be done for inbox replies and mentions

It doesn't apply to the inbox/mentions. The reasoning here is that if you get a reply or a mention, you might already have the general context of what the original discussion was about, unlike the user feed page or the search page. Additionally, sometimes you might want to just quickly reply to a reply/mention to acknowledge that you've seen it!

I'm not sure that I've tested all the edge cases everywhere, but I'm hoping that most of it gets caught during testing/nightlies.

hjiangsu commented on lib/comment/utils/navigate_comment.dart:
```diff
@@ -1,43 +1,95 @@
+            postViewMedia: postViewMedia,
+            commentView: commentView,
+            parentCommentView: parentCommentView,
+            onCommentSuccess: (CommentView commentView) {
```
Done!

hjiangsu commented:
| Before | After|
|-|-|
|![image](https://github.com/thunder-app/thunder/assets/30667958/624968df-8ff2-40dc-b2d8-eb6443a9e1c0) | ![image](https://github.com/thunder-app/thunder/assets/30667958/3354a723-5aa4-4dde-a0f4-6da0dc0aba69)|

@machinaeZER0 does this seem to be a bit better?

machinaeZER0 commented:
> Before 	After
> ![image](https://private-user-images.githubusercontent.com/30667958/323276408-624968df-8ff2-40dc-b2d8-eb6443a9e1c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNzg2OTMsIm5iZiI6MTcxMzM3ODM5MywicGF0aCI6Ii8zMDY2Nzk1OC8zMjMyNzY0MDgtNjI0OTY4ZGYtOGZmMi00MGRjLWIyZDgtZWI2NDQzYTllMWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDE4MjYzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3NjcyMTQ4ZTc2ZDJhN2EzZDRiNDU2YzUxMzhhOWVmNGZmNmM2NmVjZWM3NjFmZjFjMmQ5MWU0ZmM0ZjEzZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kx7hggSXz9xvT4nRGydE-Uwc0ScnbDd0tD65knaQVb8) 	![image](https://private-user-images.githubusercontent.com/30667958/323275855-3354a723-5aa4-4dde-a0f4-6da0dc0aba69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNzg2OTMsIm5iZiI6MTcxMzM3ODM5MywicGF0aCI6Ii8zMDY2Nzk1OC8zMjMyNzU4NTUtMzM1NGE3MjMtNWFhNC00ZGRlLWEwZjQtNmRhMGRjMGFiYTY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDE4MjYzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3NWRlZjI4MTVkZmUxYWViNzA1YmM2MTc4M2ZkZmU2ZDdiNzUzNjYzNjllMjMzYTcxNGNlNmUzMGYwOWI0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ROwZ6Gz0B_etnkktK0VUPzZ-t8lKBkcjIz61jHYkZtU)
> 
> @machinaeZER0 does this seem to be a bit better?

If it's possible to split the difference I might do that, but if that's literally one pixel's worth of difference then the original spacing is probably fine to keep as-is - probably better to not overly pad, anyway :) thank you for mocking it up, either way!
